File: 14.SuffixAutomaton.py
# A Suffix Automaton (SAM) is a powerful data structure that represents
# all substrings of a given string efficiently.
# It is the smallest deterministic finite automaton (DFA) that accepts all suffixes of a string.

# Explanation
# A Suffix Automaton compresses all substrings into a graph-like structure with states and transitions.
# Each state represents a set of substrings (end positions).
# SAM allows substring queries, longest common substring, and counting distinct substrings in O(n).

class State:
    def __init__(self):
        self.next = {}
        self.link = -1
        self.length = 0

class SuffixAutomaton:
    def __init__(self):
        self.states = [State()]
        self.last = 0

    def extend(self, c):
        cur = len(self.states)
        self.states.append(State())
        self.states[cur].length = self.states[self.last].length + 1

        p = self.last
        while p != -1 and c not in self.states[p].next:
            self.states[p].next[c] = cur
            p = self.states[p].link

        if p == -1:
            self.states[cur].link = 0
        else:
            q = self.states[p].next[c]
            if self.states[p].length + 1 == self.states[q].length:
                self.states[cur].link = q
            else:
                clone = len(self.states)
                self.states.append(State())
                self.states[clone].length = self.states[p].length + 1
                self.states[clone].next = self.states[q].next.copy()
                self.states[clone].link = self.states[q].link

                while p != -1 and self.states[p].next.get(c) == q:   # ✅ must use get()
                    self.states[p].next[c] = clone
                    p = self.states[p].link

                self.states[q].link = self.states[cur].link = clone

        self.last = cur


s = "ababa"
sam = SuffixAutomaton()
for ch in s:
    sam.extend(ch)

for i, st in enumerate(sam.states):
    print(f"State {i}: len={st.length}, link={st.link}, transitions={st.next}")

File: 14.SuffixAutomaton.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

class State {
    Map<Character, Integer> next;   // transitions from this state
    int link;                       // suffix link
    int length;                     // max length of substring for this state

    State() {
        next = new HashMap<>();
        link = -1;
        length = 0;
    }
}

class SuffixAutomaton {
    List<State> states;     // all states
    int last;               // last added state

    SuffixAutomaton() {
        states = new ArrayList<>();
        states.add(new State()); // state 0 = initial
        last = 0;
    }

    void extend(char c) {
        // Step 1: create new state
        int cur = states.size();
        states.add(new State());
        states.get(cur).length = states.get(last).length + 1;

        // Step 2: set transitions from last
        int p = last;
        while (p != -1 && !states.get(p).next.containsKey(c)) {
            states.get(p).next.put(c, cur);
            p = states.get(p).link;
        }

        // Step 3: set suffix link
        if (p == -1) {
            states.get(cur).link = 0;
        } else {
            int q = states.get(p).next.get(c);
            if (states.get(p).length + 1 == states.get(q).length) {
                states.get(cur).link = q;
            } else {
                // Clone state
                int clone = states.size();
                states.add(new State());
                states.get(clone).length = states.get(p).length + 1;
                states.get(clone).next.putAll(states.get(q).next);
                states.get(clone).link = states.get(q).link;

                while (p != -1 && states.get(p).next.get(c) == q) {
                    states.get(p).next.put(c, clone);
                    p = states.get(p).link;
                }

                states.get(q).link = states.get(cur).link = clone;
            }
        }
        last = cur; // update last
    }
}

public class Main {
    public static void main(String[] args) {
        String s = "ababa";   // Input string

        // Build SAM
        SuffixAutomaton sam = new SuffixAutomaton();
        for (char ch : s.toCharArray()) {
            sam.extend(ch);   // extend automaton with each character
        }

        // Output states
        for (int i = 0; i < sam.states.size(); i++) {
            State st = sam.states.get(i);
            System.out.println("State " + i +
                               ": len=" + st.length +
                               ", link=" + st.link +
                               ", transitions=" + st.next);
        }
    }
}
